### PR TITLE
Converted to PDO

### PR DESCRIPTION
**Pull request by jorda13456**
Mainly changed everything about the functions and it now uses PDO.

Also, you can login, register, logout etc.

Download Database here: http://www.filedropper.com/devvit1

or use appropriate table scheme:

[id][username][password][email]

id = key
username = varchar(50) characters
password = char(255) character
email = varchar(50) characters

### CODE REVIEW

File: website/connections/class.user.php [hunk 1]
```diff
@@ -1,31 +1,165 @@
...
-            return true;
-        }
-    }
-
-    $user = new user();
-
-
-?>
+<?php
+class USER{
```
Comment (jardayn) [inline website/connections/class.user.php]: User

File: website/connections/class.user.php [hunk 2]
```diff
@@ -1,31 +1,165 @@
...
+
+    private $db;
```
Comment (jardayn) [inline website/connections/class.user.php]: Why is DB still here?

File: website/connections/class.user.php [hunk 3]
```diff
@@ -1,31 +1,165 @@
...
+    public $message = array();
+
+    function __construct($db_con)
+    {
+        $this->db = $db_con;
+    }
+
+    public function register($user,$pass,$email)
```
Comment (jardayn) [inline website/connections/class.user.php]: Why is this not in UserHelper?

File: website/connections/class.user.php [hunk 4]
```diff
@@ -1,31 +1,165 @@
...
+    {
+        try
+        {
+          $protected_password = password_hash($pass, PASSWORD_DEFAULT);
+          $stmt = $this->db->prepare("INSERT INTO users(id, username,password,email)
```
Comment (jardayn) [inline website/connections/class.user.php]: You don't need to insert id. remove that shit bro

File: website/connections/class.user.php [hunk 5]
```diff
@@ -1,31 +1,165 @@
...
+          VALUES('', :username, :password, :email)");
+
+           $stmt->bindparam(":username", $user);
+           $stmt->bindparam(":password", $protected_password);
+           $stmt->bindparam(":email", $email);
+           $stmt->execute();
+
+           return $stmt;
```
Comment (jardayn) [inline website/connections/class.user.php]: return result maybe? rename the var

File: website/connections/class.user.php [hunk 6]
```diff
@@ -1,31 +1,165 @@
...
+            echo $e->getMessage();
+        }
+    }
+
+    public function login($username,$password,$email)
+    {
+       try
+       {
+          $stmt = $this->db->prepare("SELECT * FROM users WHERE username=:user OR email=:email LIMIT 1");
+          $stmt->execute(array(':user'=>$username, ':email'=>$email));
```
Comment (jardayn) [inline website/connections/class.user.php]: why are we still doing usernames? that's so 2005

Email all the way

File: website/connections/class.user.php [hunk 7]
```diff
@@ -1,31 +1,165 @@
...
+          $userRow=$stmt->fetch(PDO::FETCH_ASSOC);
+          if($stmt->rowCount() > 0)
+          {
+             if(password_verify($password, $userRow['password']))
```
Comment (jardayn) [inline website/connections/class.user.php]: Too many lines. If (){
}Else {

and not {
} 
{
} etc

File: website/connections/class.user.php [hunk 8]
```diff
@@ -1,31 +1,165 @@
...
+             }
+          }
+       }
+       catch(PDOException $e)
+       {
+           echo $e->getMessage();
+       }
+   }
+
+    public function validate()
```
Comment (jardayn) [inline website/connections/class.user.php]: Why is this here again? You have fish in your pillows.

File: website/connections/class.user.php [hunk 9]
```diff
@@ -1,31 +1,165 @@
...
+    {
+
+        if(isset($_POST['signup-btn']))
```
Comment (jardayn) [inline website/connections/class.user.php]: Why is $post in validate?

File: website/connections/class.user.php [hunk 10]
```diff
@@ -1,31 +1,165 @@
...
+        {
+        $username = $this->secure(trim($_POST['signup-username']));
+        $password = $this->secure(trim($_POST['signup-password']));
+        $email = trim($_POST['signup-email']);
+        $pass_validate = $this->secure(trim($_POST['signup-password-repeat']));
+
+        if($username=='')
```
Comment (jardayn) [inline website/connections/class.user.php]: why not write a functionthat requires it to be > X number of chars

File: website/connections/class.user.php [hunk 11]
```diff
@@ -1,31 +1,165 @@
...
+        {
+            $this->message[] = "Please enter a username";
+        }
+        else if($password=='')
```
Comment (jardayn) [inline website/connections/class.user.php]: same shit.

Also If you want to return a message array, then else if won't work the way you want. Ifs all the way

File: website/connections/class.user.php [hunk 12]
```diff
@@ -1,31 +1,165 @@
...
+
+       }
+       catch(PDOException $e)
+       {
+          echo $e->getMessage();
+       }
+       }
+     }
+
+     if(isset($_POST['login-btn']))
```
Comment (jardayn) [inline website/connections/class.user.php]: same shit. No $_POST in these functions. Pass parameters to the function bro

File: website/connections/class.user.php [hunk 13]
```diff
@@ -1,31 +1,165 @@
...
+     {
+         $username = $_POST['login-username'];
+         $password = $_POST['login-password'];
+         $email = $_POST['login-username'];
+
+         if($this->login($username,$password,$email))
+         {
+            $this->redirect('home.php');
```
Comment (jardayn) [inline website/connections/class.user.php]: Redirects have no plcae here

File: website/connections/class.user.php [hunk 14]
```diff
@@ -1,31 +1,165 @@
...
+         }
+         else
+         {
+            $this->message[] = "Wrong Information.";
+            echo $username . "<br/>";
+            echo $password . "<br/>";
+            echo $email . "<br/>";
```
Comment (jardayn) [inline website/connections/class.user.php]: This is debug code. why is it being commited uncommented?

File: website/connections/class.user.php [hunk 15]
```diff
@@ -1,31 +1,165 @@
...
+         }
+     }
+    }
+
+    public function secure($arg)
```
Comment (jardayn) [inline website/connections/class.user.php]: shitname

File: website/connections/class.user.php [hunk 16]
```diff
@@ -1,31 +1,165 @@
...
+    {
+        return mysql_real_escape_string($arg);
+    }
+
+    public function is_loggedin()
```
Comment (jardayn) [inline website/connections/class.user.php]: why is this here?

File: website/connections/class.user.php [hunk 17]
```diff
@@ -1,31 +1,165 @@
...
+    {
+        if (isset($_SESSION['user_session']))
+        {
+            return true;
+        }
+    }
+
+    public function redirect($url)
```
Comment (jardayn) [inline website/connections/class.user.php]: why is this here?

File: website/connections/class.user.php [hunk 18]
```diff
@@ -1,31 +1,165 @@
...
+    {
+        header("Location: $url");
+    }
+
+    public function logout()
```
Comment (jardayn) [inline website/connections/class.user.php]: why is this here?

File: website/connections/class.user.php [hunk 19]
```diff
@@ -1,31 +1,165 @@
...
+    {
+        session_destroy();
+        unset($_SESSION['user_session']);
+        header("Location: index.php");
```
Comment (jardayn) [inline website/connections/class.user.php]: if you have a function for redirects, why are you not using it?

File: website/connections/db.php [hunk 20]
```diff
@@ -1,61 +1,19 @@
...
-        while ($row = mysqli_fetch_array($result)){
-            $rows = $row;
-        }
-        return $rows;
-    }
+$connection = parse_ini_file('config.ini');
 
+try
+{
+    $db_con = new PDO("mysql:host={$connection['host']}; dbname={$connection['dbname']}", $connection['username'], $connection['password']);
```
Comment (jardayn) [inline website/connections/db.php]: Why is this not a class?

File: website/home.php [hunk 21]
```diff
@@ -0,0 +1,233 @@
+<?php
+include_once 'connections/db.php';
+if(!$user->is_loggedin())
+{
+ $user->redirect('index.php');
+}
+$user_id = $_SESSION['user_session'];
+$stmt = $db_con->prepare("SELECT * FROM users WHERE username=:user_id");
```
Comment (jardayn) [inline website/home.php]: is it user_id or username?

Like, user_id implies an integer. username implies a diff thing

File: website/index.php [hunk 22]
```diff
@@ -1,232 +1,231 @@
...
-        <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
-        <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4" crossorigin="anonymous"></script>
-        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script>
-    </body>
-</html>
+<?php
+include_once 'connections/db.php';
+if($user->is_loggedin())
+{
+ $user->redirect('home.php');
```
Comment (jardayn) [inline website/index.php]: /home.php

File: website/signup.php [hunk 23]
```diff
@@ -64,11 +21,11 @@
     <body>
         <div class="wrapper">
             <header class="header-top">
-                <div class="logo"><a href="#">Devvit <span class="logo-io">.io</span></a></div>
+                <div class="logo"><a href="index.php">Devvit<span class="logo-io">.io</span></a></div>
```
Comment (jardayn) [inline website/signup.php]: /index.php

paths must be absolute so you don't have a headache in the future

File: website/signup.php [hunk 24]
```diff
@@ -86,21 +43,26 @@
...
+                            <?php
+                                foreach($user->message as $message){
+                                    echo $message;
                                 }
-                            } ?>
-                            <input type="text" class="signup-form-username" name = "signup-username" placeholder="USERNAME">
-                            <input type="password" class="signup-form-password" name = "signup-password" placeholder="PASSWORD">
-                            <button type="submit" class="signup-form-submit-btn" name = "btn-signup" >Sign Up</button>
+                            ?>
+                            <input type="email" class="signup-form-input" placeholder="EMAIL ADDRESS" name="signup-email">
```
Comment (jardayn) [inline website/signup.php]: WHY ARE WE TYPING IN CAPS?

File: website/signup.php [hunk 25]
```diff
@@ -121,5 +83,6 @@
         <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
         <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4" crossorigin="anonymous"></script>
         <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script>
+        <script src="js/script.js"></script>
```
Comment (jardayn) [inline website/signup.php]: /js/script.js

File: website/connections/class.user.php (hunk 1 above)
Comment (jorda13456) [reply website/connections/class.user.php]: My Mistake.
Thanks for the catch.

File: website/connections/class.user.php (hunk 2 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I put $db there so I can access the database connection from the user class .

I put the connection into a constructor so it's automatically ran when the class user is initialized.

File: website/connections/class.user.php (hunk 3 above)
Comment (jorda13456) [reply website/connections/class.user.php]: The whole class.user.php is UserHelper.  It's made so assist the user, and all the functions inside are created to assist the user

File: website/connections/class.user.php (hunk 4 above)
Comment (jorda13456) [reply website/connections/class.user.php]: My Mistake.
Thanks for the catch.

File: website/connections/class.user.php (hunk 5 above)
Comment (jorda13456) [reply website/connections/class.user.php]: Are you saying to write:
$result = $stmt->execute();

then

return $result;

?

File: website/connections/class.user.php (hunk 6 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I think it's fine to allow the user to choose either to login.  They currently have the choice with the way I set it up

File: website/connections/class.user.php (hunk 7 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I previously did it like the way you shown:

if (){
}

only with this project I tried this new style which I found was easier to read.

File: website/connections/class.user.php (hunk 8 above)
Comment (jorda13456) [reply website/connections/class.user.php]: validate() function is called from signup.php

Inside, it checks if the user hit the register button or login button.

Depending on the option, it will run the appropriate script within validate() function.

you said not to put it on the outside(signup.php) so i moved it into it's own function.

If you have a better method, please elaborate fully

File: website/connections/class.user.php (hunk 9 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I should've user parameters instead, my fault.

File: website/connections/class.user.php (hunk 10 above)
Comment (jorda13456) [reply website/connections/class.user.php]: It checks if username is empty, not an (x) amount of characters.

Although, thank you; I forgot to add a check for the minimum amount of characters as well.

Also, that line probably should've been written as:
if(!isset($username)){
}

File: website/connections/class.user.php (hunk 11 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I'm confused by your response, it does work as I intended it.  Maybe i'm missing something though :/ Enlighten me!!

File: website/connections/class.user.php (hunk 12 above)
Comment (jorda13456) [reply website/connections/class.user.php]: My fault

File: website/connections/class.user.php (hunk 13 above)
Comment (jorda13456) [reply website/connections/class.user.php]: Explain why.

If the user verification check works, then the user is redirected to the appropriate place where I want.

Show me a better solution

File: website/connections/class.user.php (hunk 14 above)
Comment (jorda13456) [reply website/connections/class.user.php]: My Mistake.
Thanks for the catch.

File: website/connections/class.user.php (hunk 15 above)
Comment (jorda13456) [reply website/connections/class.user.php]: I can't argue with that.  I'm also not very good at choosing names, but i'm learning!

File: website/connections/class.user.php (hunk 16 above)
Comment (jorda13456) [reply website/connections/class.user.php]: Because it's related to the user.

If the user is logged in do x otherwise y.

It should be placed in there, as it's a function of the user.

Otherwise, show me a better way.

File: website/connections/class.user.php (hunk 18 above)
Comment (jorda13456) [reply website/connections/class.user.php]: A function related to the user.  Anything related to the user should be within the user class.

Otherwise, show me a better way

File: website/connections/class.user.php (hunk 19 above)
Comment (jorda13456) [reply website/connections/class.user.php]: My Mistake.
Thanks for the catch.

File: website/connections/db.php (hunk 20 above)
Comment (jorda13456) [reply website/connections/db.php]: Doesn't need to be as I chose to create new instances of existing classes within that db.php file.

It could've been but I don't see the point in this situation.

Please show me a better way, I just want to learn.

File: website/home.php (hunk 21 above)
Comment (jorda13456) [reply website/home.php]: My Mistake.
Thanks for the catch.

File: website/index.php (hunk 22 above)
Comment (jorda13456) [reply website/index.php]: Thank you, I wasn't aware of this.  I appreciate it!

File: website/signup.php (hunk 23 above)
Comment (jorda13456) [reply website/signup.php]: Thank you, I wasn't aware of this. I appreciate it!

File: website/signup.php (hunk 25 above)
Comment (jorda13456) [reply website/signup.php]: Thank you!

File: website/signup.php (hunk 24 above)
Comment (jorda13456) [reply website/signup.php]: I didn't write any of the already existing-html code, only php.

File: website/connections/class.user.php (hunk 2 above)
Comment (jardayn) [reply website/connections/class.user.php]: I need to have a chat with SirDrapple about this

File: website/connections/class.user.php (hunk 5 above)
Comment (jardayn) [reply website/connections/class.user.php]: Yup

Comment (jardayn) [reply website/connections/class.user.php]: Being explicit is important

File: website/connections/class.user.php (hunk 6 above)
Comment (jardayn) [reply website/connections/class.user.php]: Having 2 ways of logging in (username/email) tends to just confuse the user. because they have to remember what they used to register.

That's why all major websites have allowed login only through email (cos you remember them) and made Username either redundant or use it just for show

File: website/connections/class.user.php (hunk 7 above)
Comment (jardayn) [reply website/connections/class.user.php]: You get a lot of really long files. Too much empty space makes it hard to read efficiently

File: website/connections/class.user.php (hunk 8 above)
Comment (jardayn) [reply website/connections/class.user.php]: You can make a class Validator. Instantiate it. Then pass it the data you want validated and the rules you want it to be validated against.

The validator class should have defined functions for rules.

File: website/connections/class.user.php (hunk 10 above)
Comment (jardayn) [reply website/connections/class.user.php]: You can just use a validator class as disgussed above

File: website/connections/class.user.php (hunk 11 above)
Comment (jardayn) [reply website/connections/class.user.php]: http://sandbox.onlinephpfunctions.com/code/5bda3220c011db610f664ae18443e3d2f495502b

Also it's elseif, there's no space.

Strange that it works that way. Because If/elseif/else statement tend to stop executing after the first trigger (look at the example)

File: website/connections/class.user.php (hunk 13 above)
Comment (jardayn) [reply website/connections/class.user.php]: A better solution would be to call if($user->login($username,$password)){ redirect .... }else{redirect back}
 in a different file.

So your user class has easily definable tasks. Separation of concerns. Each class should have easily definable things it does.

File: website/signup.php (hunk 24 above)
Comment (jardayn) [reply website/signup.php]: Well fixing it won't make your life worse.
